configure.ac

```diff
@@ -202,6 +202,61 @@ AS_CASE([$with_gc],
     )
 AC_MSG_RESULT([$with_gc])
 
+dnl
+dnl User setting: native thread-local storage (off by default)
+dnl
+
+AC_ARG_WITH([native-tls],
+    [AS_HELP_STRING([--with-native-tls],
+        [use native thread-local storage implementation])],
+        [],
+        [with_native_tls=no])
+AC_MSG_CHECKING([whether to use native tls])
+AS_IF([test "x$with_native_tls" == "xyes"], [
+        AC_COMPILE_IFELSE(
+            [AC_LANG_SOURCE([[
+#if defined(__APPLE__) && defined(__x86_64__)
+#error FAIL
+#endif
+        ]])],
+            [
+                AC_DEFINE([USE_NATIVE_TLS], [1], [define as 1 if using native TLS])
+                AC_MSG_RESULT([yes])
+            ],
+            [
+                AC_DEFINE([USE_PTHREAD_TLS], [1], [define as 1 if using pthread_getspecific])
+                AC_MSG_RESULT([pthread])
+            ]
+        )
+        AC_RUN_IFELSE(
+            [AC_LANG_PROGRAM([[
+#include <pthread.h>
+#include <string.h>
+]],[[
+#if defined(__APPLE__) && defined(__x86_64__)
+    static char asm_code[] = {
+        // movq %gs:(,%rdi,8), %rax
+        // retq
+        0x65, 0x48, 0x8b, 0x04, 0xfd, 0x00, 0x00, 0x00, 0x00, 0xc3
+    };
+    if (memcmp((void *)pthread_getspecific, asm_code, sizeof(asm_code)) != 0) {
```

Comment:
Woo, there needs to be a MUCH bigger comment around here. I would never guess what this was doing.

What is the improvement in performance got from this? I really don't like the idea of doing this super low-level thing, particularly becuase it is super-fragile to future changes to pthread_getspecific, even if they produce an equivalent future.

Reply:
I moved this to the configure script, where it's getting its own comment.

The improvement in performance is pretty significant, runs around 20%-30% slower without or so, I seem to recall. Note that this is basically taken from the mono (open source .NET compiler/runtime) implementation (not the code, just the idea). It's a very macOS-specific hack, because Apple refuses to define that part of the ABI.

But this is also why it needs to go to the configure script in order to avoid potential fragility. I'm actually thinking about adding a couple of additional sanity checks there to ensure that it works as advertised.

Reply:
A comment linking to Mono might be a good idea, particularly because if this breaks here, the Mono people might well fix it and we can take their fix :)

Reply:
Yes, I added a reference to that. Note that it "breaking" means falling back to the slower portable version, so it would be a performance regression.

Reply:
@ChrisJefferson Reimer added a comment, please take a look to see if that is good enough for you :-)

Reply:
Could you just point to where mono does this? A link to some place in the code would be good.

Reply:
My general worry is, in the event none of the people in this PR are around when this needs fixing (which I'm sure it will, when new version of Mac OS X come out), then fixing it back up again will be a major undertaking.

Also, should we maybe have some way of noticing this is failing? (I suppose we can hope someone notices the configure test is failing, or notices the slowdown).

Reply:
You can find the matching code in mono [here](https://github.com/mono/mono/blob/6bc988b0870b7d4d56a73c13b45fe5158f81c64d/mono/mini/mini-amd64.c#L4125). The ABI has remained unchanged ever since Apple started using Intel processors, so I don't expect this to change from version to version.

Actually, it looks like while the basic code hasn't changed in forever, it's slightly different if the executable is launched from the XCode GUI. While this is unlikely to happen here, I might just as well add a check for it, too (the semantics of the two functions are necessarily the same).

And yes, it would probably be a good idea to have a way to notice when it's failing. Just not sure how to keep it from being drowned in configure spam.